File: scripts/import-obsidian.js
#!/usr/bin/env node
import fs from "node:fs";
import path from "node:path";

function getArg(name) {
  const idx = process.argv.indexOf(`--${name}`);
  if (idx === -1) return undefined;
  return process.argv[idx + 1];
}

function toTitleCase(value) {
  return value
    .split(/[\s_-]+/)
    .filter(Boolean)
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1).toLowerCase())
    .join(" ");
}

function slugify(value) {
  return value
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
}

function platformPrefix(platform) {
  const normalized = platform.toLowerCase();
  if (normalized.includes("hackthebox")) return "htb";
  if (normalized.includes("portswigger")) return "portswigger";
  return slugify(platform);
}

function normalizeImagePath(rawTarget) {
  const cleaned = rawTarget.trim().replace(/^["']|["']$/g, "");
  if (
    cleaned.startsWith("http://") ||
    cleaned.startsWith("https://") ||
    cleaned.startsWith("data:") ||
    cleaned.startsWith("#")
  ) {
    return cleaned;
  }

  const noDotPrefix = cleaned.replace(/^\.\//, "");
  if (noDotPrefix.startsWith("images/")) return `./${noDotPrefix}`;

  if (!noDotPrefix.includes("/")) return `./images/${noDotPrefix}`;

  return `./${noDotPrefix}`;
}

function fixImageLinks(content) {
  const wikiFixed = content.replace(/!\[\[([^\]|]+)(?:\|([^\]]+))?\]\]/g, (_, target, alt) => {
    const finalTarget = normalizeImagePath(target);
    const finalAlt = (alt ?? path.basename(target)).replace(/\.[^.]+$/, "");
    return `![${finalAlt}](${finalTarget})`;
  });

  return wikiFixed.replace(/!\[([^\]]*)\]\(([^)]+)\)/g, (_, alt, target) => {
    return `![${alt}](${normalizeImagePath(target)})`;
  });
}

function ensureDir(targetPath) {
  if (!fs.existsSync(targetPath)) fs.mkdirSync(targetPath, { recursive: true });
}

function parseSourceMetadata(sourceDir) {
  const parts = sourceDir.split(path.sep);
  const challenge = parts[parts.length - 1];
  const category = parts[parts.length - 2];
  const challengesFolder = parts[parts.length - 3] ?? "";
  const platform = challengesFolder.replace(/\s*challenges\s*$/i, "") || "Unknown";
  return {
    title: challenge,
    categoryRaw: category,
    category: toTitleCase(category),
    platform
  };
}

function frontmatterBlock(meta, date) {
  return [
    "---",
    `title: "${meta.title}"`,
    `date: ${date}`,
    `platform: "${meta.platform}"`,
    `category: "${meta.category}"`,
    'difficulty: "Medium"',
    `tags: ["${slugify(meta.categoryRaw)}", "${slugify(meta.platform)}"]`,
    "draft: false",
    "---",
    ""
  ].join("\n");
}

function findChallengeDirs(rootDir) {
  const found = [];
  const stack = [rootDir];

  while (stack.length > 0) {
    const current = stack.pop();
    if (!current || !fs.existsSync(current)) continue;

    const readmePath = path.join(current, "README.md");
    if (fs.existsSync(readmePath)) {
      found.push(current);
      continue;
    }

    const entries = fs.readdirSync(current, { withFileTypes: true });
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      if (entry.name.startsWith(".")) continue;
      stack.push(path.join(current, entry.name));
    }
  }

  return found.sort((a, b) => a.localeCompare(b));
}

function importOne(sourceDir, { outRoot, date, overwrite }) {
  const absSource = path.resolve(sourceDir);
  const readmePath = path.join(absSource, "README.md");
  const imagesPath = path.join(absSource, "images");

  if (!fs.existsSync(readmePath)) {
    throw new Error(`README.md not found at: ${readmePath}`);
  }

  const meta = parseSourceMetadata(absSource);
  const bundleSlug = `${platformPrefix(meta.platform)}-${slugify(meta.categoryRaw)}-${slugify(meta.title)}`;
  const absOutRoot = path.resolve(outRoot);
  const bundleDir = path.join(absOutRoot, bundleSlug);
  const indexPath = path.join(bundleDir, "index.mdx");
  const outImagesPath = path.join(bundleDir, "images");

  if (fs.existsSync(bundleDir)) {
    if (!overwrite) {
      return { skipped: true, source: absSource, bundle: bundleDir };
    }
    fs.rmSync(bundleDir, { recursive: true, force: true });
  }

  ensureDir(bundleDir);

  if (fs.existsSync(imagesPath)) {
    fs.cpSync(imagesPath, outImagesPath, { recursive: true });
  }

  let content = fs.readFileSync(readmePath, "utf8");
  content = fixImageLinks(content);

  if (!content.trimStart().startsWith("---")) {
    content = `${frontmatterBlock(meta, date)}\n${content}`;
  }

  fs.writeFileSync(indexPath, content, "utf8");
  return { skipped: false, source: absSource, bundle: bundleDir, index: indexPath };
}

function main() {
  const sourceDir = getArg("source");
  const outRoot = getArg("out") ?? "src/content/ctf";
  const date = getArg("date") ?? new Date().toISOString().slice(0, 10);
  const overwrite = process.argv.includes("--overwrite");
  const bulk = process.argv.includes("--bulk");

  if (!sourceDir) {
    console.error("Missing required argument: --source <path-to-challenge-folder-or-root>");
    process.exit(1);
  }

  const absSource = path.resolve(sourceDir);
  if (!fs.existsSync(absSource)) {
    console.error(`Source path does not exist: ${absSource}`);
    process.exit(1);
  }

  if (!bulk) {
    const result = importOne(absSource, { outRoot, date, overwrite });
    if (result.skipped) {
      console.log(`Skipped (exists): ${result.bundle}`);
      return;
    }
    console.log(`Imported: ${result.source}`);
    console.log(`Bundle: ${result.bundle}`);
    console.log(`Entry: ${result.index}`);
    return;
  }

  const challengeDirs = findChallengeDirs(absSource);
  if (challengeDirs.length === 0) {
    console.error(`No README.md challenge folders found under: ${absSource}`);
    process.exit(1);
  }

  let imported = 0;
  let skipped = 0;
  for (const dir of challengeDirs) {
    try {
      const result = importOne(dir, { outRoot, date, overwrite });
      if (result.skipped) {
        skipped += 1;
        console.log(`Skipped: ${result.bundle}`);
      } else {
        imported += 1;
        console.log(`Imported: ${result.bundle}`);
      }
    } catch (error) {
      console.error(`Failed: ${dir}`);
      console.error(error instanceof Error ? error.message : String(error));
    }
  }

  console.log(`Done. Imported: ${imported}, Skipped: ${skipped}, Total: ${challengeDirs.length}`);
}

main();

File: scripts/import_obsidian.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import re
import shutil
from datetime import date
from pathlib import Path


def to_title_case(value: str) -> str:
    parts = re.split(r"[\s_-]+", value.strip())
    parts = [p for p in parts if p]
    return " ".join(p[:1].upper() + p[1:].lower() for p in parts)


def slugify(value: str) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", value.lower())
    return re.sub(r"^-+|-+$", "", slug)


def platform_prefix(platform: str) -> str:
    normalized = platform.lower()
    if "hackthebox" in normalized:
        return "htb"
    if "portswigger" in normalized:
        return "portswigger"
    return slugify(platform)


def normalize_image_path(raw_target: str) -> str:
    cleaned = raw_target.strip().strip("'\"")
    if cleaned.startswith(("http://", "https://", "data:", "#")):
        return cleaned

    no_dot_prefix = re.sub(r"^\./", "", cleaned)
    if no_dot_prefix.startswith("images/"):
        return f"./{no_dot_prefix}"
    if "/" not in no_dot_prefix:
        return f"./images/{no_dot_prefix}"
    return f"./{no_dot_prefix}"


def fix_image_links(content: str) -> str:
    wiki_pattern = re.compile(r"!\[\[([^\]|]+)(?:\|([^\]]+))?\]\]")
    md_pattern = re.compile(r"!\[([^\]]*)\]\(([^)]+)\)")

    def replace_wiki(match: re.Match[str]) -> str:
        target = match.group(1)
        alt = match.group(2) or Path(target).stem
        return f"![{alt}]({normalize_image_path(target)})"

    def replace_md(match: re.Match[str]) -> str:
        alt = match.group(1)
        target = match.group(2)
        return f"![{alt}]({normalize_image_path(target)})"

    content = wiki_pattern.sub(replace_wiki, content)
    content = md_pattern.sub(replace_md, content)
    return content


def parse_source_metadata(source_dir: Path) -> dict[str, str]:
    challenge = source_dir.name
    category = source_dir.parent.name
    challenges_folder = source_dir.parent.parent.name if source_dir.parent.parent else ""
    platform = re.sub(r"\s*challenges\s*$", "", challenges_folder, flags=re.IGNORECASE) or "Unknown"
    return {
        "title": challenge,
        "category_raw": category,
        "category": to_title_case(category),
        "platform": platform,
    }


def frontmatter_block(meta: dict[str, str], post_date: str) -> str:
    return "\n".join(
        [
            "---",
            f'title: "{meta["title"]}"',
            f"date: {post_date}",
            f'platform: "{meta["platform"]}"',
            f'category: "{meta["category"]}"',
            'difficulty: "Medium"',
            f'tags: ["{slugify(meta["category_raw"])}", "{slugify(meta["platform"])}"]',
            "draft: false",
            "---",
            "",
        ]
    )


def find_challenge_dirs(root_dir: Path) -> list[Path]:
    found: list[Path] = []
    for readme in root_dir.rglob("README.md"):
        found.append(readme.parent)
    return sorted(found)


def import_one(source_dir: Path, out_root: Path, post_date: str, overwrite: bool) -> tuple[bool, Path]:
    readme_path = source_dir / "README.md"
    images_path = source_dir / "images"
    if not readme_path.exists():
        raise FileNotFoundError(f"README.md not found at: {readme_path}")

    meta = parse_source_metadata(source_dir)
    bundle_slug = f'{platform_prefix(meta["platform"])}-{slugify(meta["category_raw"])}-{slugify(meta["title"])}'
    bundle_dir = out_root / bundle_slug
    index_path = bundle_dir / "index.mdx"

    if bundle_dir.exists():
        if not overwrite:
            return False, bundle_dir
        shutil.rmtree(bundle_dir)

    bundle_dir.mkdir(parents=True, exist_ok=True)

    if images_path.exists():
        shutil.copytree(images_path, bundle_dir / "images")

    content = readme_path.read_text(encoding="utf-8")
    content = fix_image_links(content)

    if not content.lstrip().startswith("---"):
        content = f"{frontmatter_block(meta, post_date)}\n{content}"

    index_path.write_text(content, encoding="utf-8")
    return True, bundle_dir


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Import Obsidian CTF writeups into Astro content page bundles."
    )
    parser.add_argument("--source", required=True, help="Path to challenge folder or CTF root folder")
    parser.add_argument("--out", default="src/content/ctf", help="Output root for Astro CTF content")
    parser.add_argument("--date", default=str(date.today()), help="Frontmatter date (YYYY-MM-DD)")
    parser.add_argument("--overwrite", action="store_true", help="Overwrite existing destination bundles")
    parser.add_argument("--bulk", action="store_true", help="Import all README.md challenge folders recursively")
    args = parser.parse_args()

    source = Path(args.source).resolve()
    out_root = Path(args.out).resolve()

    if not source.exists():
        raise SystemExit(f"Source path does not exist: {source}")

    if args.bulk:
        challenge_dirs = find_challenge_dirs(source)
        if not challenge_dirs:
            raise SystemExit(f"No README.md challenge folders found under: {source}")

        imported = 0
        skipped = 0
        for challenge_dir in challenge_dirs:
            try:
                did_import, bundle_dir = import_one(
                    challenge_dir, out_root, args.date, overwrite=args.overwrite
                )
                if did_import:
                    imported += 1
                    print(f"Imported: {bundle_dir}")
                else:
                    skipped += 1
                    print(f"Skipped: {bundle_dir}")
            except Exception as exc:
                print(f"Failed: {challenge_dir}")
                print(str(exc))

        print(f"Done. Imported: {imported}, Skipped: {skipped}, Total: {len(challenge_dirs)}")
        return

    did_import, bundle_dir = import_one(source, out_root, args.date, overwrite=args.overwrite)
    if did_import:
        print(f"Imported: {bundle_dir}")
    else:
        print(f"Skipped (exists): {bundle_dir}")


if __name__ == "__main__":
    main()
